### ISSUE TITLE
Use-case for isReadDataAvailable and isWriteDataRequired?

### ISSUE THREAD

**Issue by BenjaminRodenberg** [precice/precice#1223]
I am not sure that the functions `isReadDataAvailable` and `isWriteDataRequired` are actually needed. Removing them is clearly breaking, but would also help us to simplify the codebase and the API. This issue is meant for discussing the use-case of these function and to find places where these functions are actually used and needed. I would like to find out whether there are actually any adapters or other user codes that really depend on these functions?

[In our documentation](https://precice.org/couple-your-code-timestep-sizes.html) we say the following about the use-case of these two functions:

> If a participant subcycles it is actually not necessary to write data to or read data from preCICE. To avoid unnecessary calls, preCICE offers two optional helper functions:
>
> bool isReadDataAvailable () const;
> bool isWriteDataRequired (double computedTimestepLength) const;

With the current developments around waveform relaxation and subcycling, I don't think that this use-case is relevant anymore:

* If an interpolant is created by preCICE to allow for subcycling, reading data is allowed at any point in time. preCICE cannot know whether the data the user is requesting is new data or the user performs unnecessary read data calls. Therefore, I would suggest to remove `isReadDataAvailable` and let the user decide whether calling a `readData` function makes sense. (these are the current plans for the "read direction" of waveform relaxation, see https://github.com/precice/precice/issues/1174)
* For creating an interpolant on a window, any kind of data provided by the user might be useful. Currently this "write direction" is still in a very early stage, but according to my current plans `isWriteDataRequired` would always return `true`.

Another important point: The two functions are barely used in tests (outside of the tests that I have implemented for waveform relaxation). This shows from my perspective again that these two functions are not really needed.

**Comment by BenjaminRodenberg** [comment]
To test this further, I removed the functions on a branch and did not run into any major problems. See c5b63562962c4d488466a653d71307254a240185 and 5d452d677577212e463d5a153ed858a8ac35829a.

**Comment by fsimonis** [comment]
Does the user actually need these functions?

If the user chooses a dt different than the return value of advance (or initialize) in the code, then he consciously chooses to subcycle and doesn't need to write data.
If the user chooses the same dt as the return value of advance or initialize, then she consciously finishes the time-window and needs to read new data.

So, the user actually already knows when to read and write.

Am I missing something?

**Comment by uekerman** [comment]
In principle correct. Two expections:

* With the new WI, they might "subcycle" (use a smaller dt than the remaining window size), but data is still required / available.
* For some codes, you might want to call `isReadDataAvailable` before computing the adaptive timestep size.

**Comment by BenjaminRodenberg** [comment]
Maybe I'm missing something, but I don't see the actual effect of these exceptions. 

> * With the new WI, they might "subcycle" (use a smaller dt than the remaining window size), but data is still required / available.

If you are subcycling there are two possibilities:

* I'm using WI. Then data will always be available and `isReadDataAvailable` will always return true, because there is an interpolant and preCICE cannot know where I want to sample it.
* I'm not using WI. Then `isReadDataAvailable` will only return true, if I'm entering a new window or iteration. But this does not have an observable effect: I either keep the old data, because I don't read data or I just get the old data from preCICE when I read.

There is, of course, the performance argument, but so far it looks to me very hypothetical. If we see that performance is really an issue here that the user cannot solve alone, we could still re-implement the function at a later point in time (non-breaking feature release).

> * For some codes, you might want to call `isReadDataAvailable` before computing the adaptive timestep size.

I think you could also fix this with `isTimeWindowComplete`. But without an actual example this is really difficult to evaluate.

But let's maybe discuss this personally, before hypothesizing too much here. If we can come up with an actual integration test where there is an observable effect this would be valuable - also on the long run.

**Comment by uekerman** [comment]
> There is, of course, the performance argument, but so far it looks to me very hypothetical.

It is only the performance argument. That's the only use case of these functions. Imagine that you have to compute an expensive mapping operation in an adapter each time you read data from preCICE.

> > For some codes, you might want to call isReadDataAvailable before computing the adaptive timestep size.

> I think you could also fix this with isTimeWindowComplete. But without an actual example this is really difficult to evaluate.

No, I don't think so.
Imagine the solver computes the adaptive timestep size based on some stability (CFL-like) criterion. If you read velocity values from preCICE and you want to incorporate them, you have to read first and then you can only compute the timestep size.

**Comment by uekerman** [comment]
Let's discuss at the next coding days how we treat such features with v3.0

**Comment by BenjaminRodenberg** [comment]
I will quickly summarize a discussion with @uekerman about the implementation of `isReadDataAvailable` from #1187:

My implementation was the following:

```
bool SolverInterfaceImpl::isReadDataAvailable() const
{
  PRECICE_TRACE();
  PRECICE_CHECK(_state != State::Constructed, "initialize() has to be called before isReadDataAvailable().");
  PRECICE_CHECK(_state != State::Finalized, "isReadDataAvailable() cannot be called after finalize().");
  bool available = _couplingScheme->hasDataBeenReceived();
  available |= (_allowsExperimental && _couplingScheme->hasInitialDataBeenReceived()); // if waveform relaxation is allowed, we always return true as soon as initialData is available.
  return available;
}
```

Main problems:

* Don't use `_allowsExperimental`, because it is the outer safeguard.
* Using `_allowsExperimental` is not sustainable, because we have to reconsider the implementation, when the API leaves experimental stage.

Solution:

* Check the waveform order of read data. If any read data of the participant is of order >= 1, then data is always available. If order == 0, then the behavior unchanged (even with the experimental API)

Is still see some problems here. These problems are not critical for #1187, but they might help us to conclude here. Therefore, I will summarize them:

* If waveform order is 0 I don't see a difference to waveform order > 0. The user might still want to sample the interpolant, even if the data will always be the same due to constant interpolation. This behavior would also be consistent with the behavior for order > 0. Differentiating here makes the behavior of the interface overly complex (keeping the unclear performance gain in mind).
* Waveform order might be different for different pieces of read data. If we want to make it right, `isReadDataAvailable` should only return `true`, if the user want to read data where order > 0. We should not check for the maximum waveform order of all read data, but for the order of a specific piece of read data. This means we would have to extend the API to accept an argument: `isReadDataAvailable(int dataID)`. This makes the API overly complex and therefore I did not implement this in #1187.

I implemented the proposed solution in 9aeeed7b9c66e94b20f3e2c46cbb781b1ca0cb4c. Taking a look at the commit and the changes in the tests might be helpful for better understanding.

**Comment by DavidSCN** [comment]
We use this functions in the [openfoam-adapter](https://github.com/precice/openfoam-adapter/blob/bf38c42a197ca9b85b33cf16575010653cea53db/Interface.C#L375) as well as the [dealii-adapter](https://github.com/precice/dealii-adapter/blob/005933d6b45f885a64aee7ce597a3d7d81d54aff/include/adapter/adapter.h#L374-L398)

**Comment by BenjaminRodenberg** [comment]
#1362 is merged into `develop-v3.0.0`.
